my/core/query_range.py

```diff
@@ -0,0 +1,526 @@
+"""
+An extension of the my.core.query.select function, allowing you to specify
+a type or key to filter the range by -- this creates a filter function
+given those values, coercing values on the iterable, returning you a
+filtered iterator
+
+See the select_range function below
+"""
+
+import re
+import time
+from functools import lru_cache
+from datetime import datetime, timedelta, date
+from typing import Callable, Iterator, NamedTuple, Optional, Any, Type
+
+import more_itertools
+
+from .query import (
+    QueryException,
+    select,
+    OrderFunc,
+    Where,
+    _handle_generate_order_by,
+    ET,
+)
+
+fromisoformat: Callable[[str], datetime]
+import sys
+if sys.version_info[:2] >= (3, 7):
+    # prevent mypy on py3.6 from complaining...
+    fromisoformat_real = datetime.fromisoformat
+    fromisoformat = fromisoformat_real
+else:
+    from .py37 import fromisoformat
+
+
+timedelta_regex = re.compile(r"^((?P<weeks>[\.\d]+?)w)?((?P<days>[\.\d]+?)d)?((?P<hours>[\.\d]+?)h)?((?P<minutes>[\.\d]+?)m)?((?P<seconds>[\.\d]+?)s)?$")
+
+
+# https://stackoverflow.com/a/51916936
+def parse_timedelta_string(timedelta_str: str) -> timedelta:
+    """
+    This uses a syntax similar to the 'GNU sleep' command
+    e.g.: 1w5d5h10m50s means '1 week, 5 days, 5 hours, 10 minutes, 50 seconds'
+    """
+    parts = timedelta_regex.match(timedelta_str)
+    if parts is None:
+        raise ValueError(f"Could not parse time duration from {timedelta_str}.\nValid examples: '8h', '1w2d8h5m20s', '2m4s'")
+    time_params = {name: float(param) for name, param in parts.groupdict().items() if param}
+    return timedelta(**time_params)  # type: ignore[arg-type]
+
+
+def parse_timedelta_float(timedelta_str: str) -> float:
+    return parse_timedelta_string(timedelta_str).total_seconds()
+
+
+def parse_datetime_float(date_str: str) -> float:
+    """
+    parses multiple possible representations of a datetime
+    into a float, else raises a QueryException
+
+    the query_cli interface compares floats instead of timestamps
+    when comparing datetimes since handling it is unknown
+    whether the sources the user is selecting from is tz-aware
+    or not (or perhaps a mix of both?)
+    """
+    ds = date_str.strip()
+    # special case
+    if ds == "now":
+        return time.time()
+    # epoch timestamp
+    try:
+        # also handles epoch timestamps as integers
+        ds_float = float(ds)
+        # convert to make sure its a valid datetime
+        datetime.fromtimestamp(ds_float)
+    except ValueError:
+        pass
+    else:
+        return ds_float
+    try:
+        # isoformat - default format when you call str() on datetime
+        return fromisoformat(ds).timestamp()
+    except ValueError:
+        raise QueryException(f"Was not able to parse {ds} into a datetime")
+    # todo: add isoparse from my.core.common? not sure how it handles
```

Comment:
fromisoformat should support timezones.. although it wants a UTC offset like `+04:00`